File: nodejs-api/src/api/handlers/results/updateResult/updateResultHandler.js
'use strict';

const AWS = require('aws-sdk')
const db = new AWS.DynamoDB.DocumentClient({})

const { POLLS_TABLE_NAME, RESULTS_ID } = process.env


const response = (statusCode, message) => {
    return {
        statusCode: statusCode,
        body: JSON.stringify(message)
    }
}

const updateResult = async (event) => {
    let PK = event.pathParameters.id
    PK = `${RESULTS_ID}#${PK}`

    const result = JSON.parse(event.body)

    const SK = result.SK
    
    let toUpdatePropers = []
    let ExpressionAttributeNames = {}
    let ExpressionAttributeValues = {}

    for (const property in result) {

      if (property === 'PK' || property === 'SK') {
          continue
      }
      
      toUpdatePropers.push(`#${property} = :${property}`)
      ExpressionAttributeNames['#' + property] = property 
      
      let resultProperty = result[property]
      
      if(property === 'groups') {
        resultProperty = JSON.stringify(resultProperty)
      }

      ExpressionAttributeValues[':' + property] = resultProperty
    }

    const UpdateExpression = `set ${toUpdatePropers.join(',')}`

    console.log('UpdateExpression: ', UpdateExpression)
    console.log('ExpressionAttributeNames: ', ExpressionAttributeNames)
    console.log('ExpressionAttributeValues: ', ExpressionAttributeValues)

    console.log('PK: ', PK)
    console.log('SK: ', SK)
    
    const params = {
      TableName: POLLS_TABLE_NAME,
      Key: {
        PK,
        SK
      },
      ConditionExpression: 'attribute_exists(PK) AND attribute_exists(SK)',
      
      UpdateExpression,
      ExpressionAttributeNames,
      ExpressionAttributeValues,
      ReturnValues: 'ALL_NEW'
    }
  
    try {
      let updateResult = await db.update(params).promise()
      updateResult = updateResult.Attributes

      const groups = updateResult?.groups

      if (groups) {
        updateResult['groups'] = JSON.parse(groups)
      }

      return response(200, updateResult)
    } catch (err) {
      console.log('err: ', err)
      return response(500, 'Error updating the Result')
    }
}

module.exports = {
    updateResult
}

File: python-api/src/api/handlers/results/updateResult/updateResultHandler.py
try:
    import unzip_requirements
except ImportError:
    pass

import boto3
import os
import simplejson as json

dynamodb_client = boto3.resource('dynamodb')
TABLE_NAME = os.environ['POLLS_TABLE_NAME']
RESULTS_ID = os.environ['RESULTS_ID']

def create_response(code, message):
  return {
    'statusCode': code,
    'headers': {'Content-Type': 'application/json'},
    'body': json.dumps(message)
  }

def get_update_params(body):
    update_expression = ["set "]
    update_values = dict()

    for key, val in body.items():
        if key == 'PK' or key == 'SK':
            continue
        update_expression.append(f" {key} = :{key},")
        update_values[f":{key}"] = val

    return "".join(update_expression)[:-1], update_values

def updateResult(event, context):

    PK = event['pathParameters']['id']
    result = json.loads(event['body'])

    SK = result['SK']

    updateExpression, expressionAttributesValues = get_update_params(result)
    
    params = {
        'Key': {
            'PK': RESULTS_ID + '#' + PK,
            'SK': SK,
        },
        'ConditionExpression': 'attribute_exists(PK) AND attribute_exists(SK)',
        'UpdateExpression': updateExpression,
        'ExpressionAttributeValues': expressionAttributesValues,
        'ReturnValues': 'ALL_NEW'
    }

    try:
        table = dynamodb_client.Table(TABLE_NAME)
        response = table.update_item(**params)
    except Exception as ex:
        print('Ex: ', ex)
        return create_response(500, 'Error updating the Poll')

    print('Response: ', response)

    if response['ResponseMetadata']['HTTPStatusCode'] != 200:
      return create_response(500, 'Error updating the Poll')

    response = response['Attributes']

    return create_response(200, response)
